core/include/drivers/rstctrl.h

```diff
@@ -0,0 +1,217 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Linaro Limited
+ *
+ * Self contained reset controller framework in rstctrl.h
+ * A reset controller exposes reset level controls to a
+ * platform.
+ */
+
+#ifndef __DRIVERS_RSTCTRL_H
+#define __DRIVERS_RSTCTRL_H
+
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+
+struct rstctrl;
+
+struct rstctrl_ops {
+	/*
+	 * Operators on reset control(s) exposed by a reset controller
+	 *
+	 * assert_level: Assert reset level on control with a timeout hint
+	 * deassert_level: Deassert reset level on control with a timeout hint
+	 * get_name: Get a string name for the controller, or NULL is none
+	 *
+	 * @rstctrl: Reset control reference
+	 * @id: Identifier for the reset level control in the reset controller
+	 * @to_ms: Timeout in microsencond or RSTCTRL_NO_TIMEOUT, may be ignored
+	 * by reset controller.
+	 * Return a TEE_Result compliant code
+	 */
+	TEE_Result (*assert_level)(struct rstctrl *rstctrl, unsigned int to_us);
+	TEE_Result (*deassert_level)(struct rstctrl *rstctrl,
+				     unsigned int to_us);
+	const char *(*get_name)(struct rstctrl *rstctrl);
+};
+
+/*
+ * struct rstctrl - Instance of a control exposed by a reset controller
+ * @ops: Operators of the reset controller
+ * @priv: Reset control private data (unsigned value or memory pointer)
+ * @exclusive: Set when a consumer has exclusive control on the reset level
+ */
+struct rstctrl {
+	const struct rstctrl_ops *ops;
+	uintptr_t priv;
+	bool exclusive;
+};
+
+/**
+ * RSTCTRL_DECLARE - Declare a reset controller driver with a single
+ * DTT compatible string.
+ *
+ * @__name: Reset controller driver name
+ * @__compat: Compatible string
+ * @__probe: Reset controller probe function
+ */
+#define RSTCTRL_DT_DECLARE(__name, __compat, __probe) \
+	static const struct dt_device_match __name ## _match_table[] = { \
+		{ .compatible = __compat }, \
+		{ } \
+	}; \
+	const struct dt_driver __name ## _dt_driver __dt_driver = { \
+		.name = # __name, \
+		.type = DT_DRIVER_RSTCTRL, \
+		.match_table = __name ## _match_table, \
+		.probe = __probe, \
+	}
+
+#define RSTCTRL_OPS_IS_VALID(_ops) \
```

Comment:
Any reason why this is a macro instead of an inline function? The latter should take care of the checkpatch issue.

Reply:
Right, I'll move to an inline function. (I should have done it seen checkpatch report)